File: src-tauri/src/bms/work.rs
use std::cmp::Ordering;
use std::collections::HashMap;

/// Extract common work name from multiple titles (improved version)
///
/// # Parameters
/// - `titles`: slice containing multiple titles
/// - `remove_unclosed_pair`: whether to remove unclosed bracket pairs (default: true)
/// - `remove_tailing_sign_list`: list of trailing symbols to remove (default: empty)
///
/// # Returns
/// Extracted common work name (post-processed)
#[must_use]
pub fn extract_work_name(
    titles: &[&str],
    remove_unclosed_pair: bool,
    remove_tailing_sign_list: &[&str],
) -> String {
    // If title list is empty, return empty string directly
    if titles.is_empty() {
        return "[!!! EMPTY !!!]".to_string();
    }

    // Use HashMap to count occurrences of all possible prefixes
    let mut prefix_counts: HashMap<String, usize> = HashMap::new();

    // Iterate through each title
    for &title in titles {
        // Generate all prefixes of the title (by character length)
        for i in 1..=title.chars().count() {
            // Get prefix composed of first i characters
            let prefix: String = title.chars().take(i).collect();
            // Update prefix count
            *prefix_counts.entry(prefix).or_insert(0) += 1;
        }
    }

    // Find maximum occurrence count
    let max_count = *prefix_counts.values().max().unwrap_or(&0);

    // Filter candidate prefixes: occurrence count exceeds 2/3 of maximum count
    let mut candidates: Vec<(String, usize)> = prefix_counts
        .into_iter()
        .filter(|(_, count)| *count as f32 >= max_count as f32 * 0.67)
        .collect();

    // Sort candidate prefixes (priority: length descending, then count descending, finally lexicographic ascending)
    candidates.sort_by(|a, b| {
        // 1. Sort by length descending
        let len_cmp = b.0.len().cmp(&a.0.len());
        if len_cmp != Ordering::Equal {
            return len_cmp;
        }

        // 2. Sort by occurrence count descending
        let count_cmp = b.1.cmp(&a.1);
        if count_cmp != Ordering::Equal {
            return count_cmp;
        }

        // 3. Sort by lexicographic ascending
        a.0.cmp(&b.0)
    });

    // Extract optimal candidate (take first if exists, otherwise empty string)
    let best_candidate = candidates
        .first()
        .map(|(s, _)| s.clone())
        .unwrap_or_default();

    // Post-process the optimal candidate
    extract_work_name_post_process(
        &best_candidate,
        remove_unclosed_pair,
        remove_tailing_sign_list,
    )
}

/// Work name post-processing function: remove unclosed brackets and trailing symbols
///
/// # Parameters
/// - `s`: original string
/// - `remove_unclosed_pair`: whether to process unclosed brackets
/// - `remove_tailing_sign_list`: list of trailing symbols to remove
///
/// # Returns
/// Processed string
fn extract_work_name_post_process(
    s: &str,
    remove_unclosed_pair: bool,
    remove_tailing_sign_list: &[&str],
) -> String {
    // First remove leading and trailing whitespace
    let mut result = s.trim().to_string();

    // Define supported bracket pairs (including full-width and half-width)
    const PAIRS: [(char, char); 7] = [
        ('(', ')'),
        ('[', ']'),
        ('{', '}'),
        ('（', '）'),
        ('［', '］'),
        ('｛', '｝'),
        ('【', '】'),
    ];

    // Loop until no changes
    loop {
        let mut changed = false;

        // Process unclosed brackets
        if remove_unclosed_pair {
            // Store bracket state (bracket character + byte position)
            let mut stack: Vec<(char, usize)> = Vec::new();

            // Iterate through each character of the string (with byte index)
            for (byte_idx, c) in result.char_indices() {
                // Check if it's an opening bracket
                if PAIRS.iter().any(|&(open, _)| c == open) {
                    stack.push((c, byte_idx));
                }
                // Check if it's a closing bracket
                else if let Some(&(last_open, _)) = stack.last() {
                    // Find matching closing bracket
                    if let Some(&(_, close)) = PAIRS.iter().find(|&&(open, _)| open == last_open)
                        && c == close
                    {
                        stack.pop();
                    }
                }
            }

            // If there are unclosed brackets
            if let Some(&(_, unmatched_pos)) = stack.last() {
                // Truncate to the position of the first unmatched bracket
                result.truncate(unmatched_pos);
                // Remove trailing whitespace after truncation
                result = result.trim_end().to_string();
                changed = true;
            }
        }

        // Process trailing symbols
        for &sign in remove_tailing_sign_list {
            if result.ends_with(sign) {
                // Remove matching trailing symbol
                result.truncate(result.len() - sign.len());
                // Remove trailing whitespace
                result = result.trim_end().to_string();
                changed = true;
                // Only remove one symbol at a time, then recheck
                break;
            }
        }

        // End processing if no changes
        if !changed {
            break;
        }
    }

    result
}

File: src/lib/utils/bms/work.ts
/**
 * BMS 工作名称提取模块
 * 从 Python 代码迁移：legacy/bms/work.py
 */

/**
 * 括号对定义
 */
const BRACKET_PAIRS: [string, string][] = [
  ['(', ')'],
  ['[', ']'],
  ['{', '}'],
  ['（', '）'],
  ['［', '］'],
  ['｛', '｝'],
  ['【', '】'],
];

/**
 * 从多个 BMS 标题中提取共同的作品名（改进版）
 *
 * @param titles - 包含多个 BMS 标题的列表
 * @param removeUnlosedPair - 是否移除未闭合括号及其后续内容
 * @param removeTailingSignList - 要移除的尾部符号列表
 * @returns 提取出的共同作品名（经过后处理）
 *
 * @command
 * @category bms
 * @dangerous false
 * @name 提取作品名称
 * @description 从多个 BMS 标题中提取共同的作品名
 * @frontend true
 */
export function extractWorkName(
  titles: string[],
  removeUnlosedPair: boolean = true,
  removeTailingSignList?: string[]
): string {
  if (titles.length === 0) {
    return '';
  }

  // 统计所有可能前缀的出现次数
  const prefixCounts = new Map<string, number>();

  for (const title of titles) {
    for (let i = 1; i <= title.length; i++) {
      const prefix = title.substring(0, i);
      const count = prefixCounts.get(prefix) || 0;
      prefixCounts.set(prefix, count + 1);
    }
  }

  if (prefixCounts.size === 0) {
    return '';
  }

  // 当只有一个标题时，没有共同前缀，直接返回空字符串
  if (titles.length <= 1) {
    return '';
  }

  // 找到最大出现次数
  const maxCount = Math.max(...prefixCounts.values());

  // 筛选出超过 2/3 次数的前缀
  const threshold = maxCount * 0.67;
  const candidates: [string, number][] = [];
  for (const [prefix, count] of prefixCounts.entries()) {
    if (count >= threshold) {
      candidates.push([prefix, count]);
    }
  }

  if (candidates.length === 0) {
    return '';
  }

  // 排序规则：优先长度降序，其次次数降序，最后字典序升序
  candidates.sort((a, b) => {
    if (a[0].length !== b[0].length) {
      return b[0].length - a[0].length; // 长度降序
    }
    if (a[1] !== b[1]) {
      return b[1] - a[1]; // 次数降序
    }
    return a[0].localeCompare(b[0]); // 字典序升序
  });

  // 提取最优候选
  const bestCandidate = candidates[0][0];

  // 后处理：移除未闭合括号及其后续内容
  return extractWorkNamePostProcess(bestCandidate, removeUnlosedPair, removeTailingSignList);
}

/**
 * 后处理函数：移除未闭合括号及其后续内容
 *
 * @param s - 原始字符串
 * @param removeUnlosedPair - 是否移除未闭合括号
 * @param removeTailingSignList - 要移除的尾部符号列表
 * @returns 处理后的字符串
 */
function extractWorkNamePostProcess(
  s: string,
  removeUnlosedPair: boolean,
  removeTailingSignList?: string[]
): string {
  // 清除前后空格
  let result = s.trim();

  if (removeTailingSignList === undefined) {
    removeTailingSignList = [];
  }

  while (true) {
    let triggered = false;

    if (removeUnlosedPair) {
      // 使用栈记录括号状态
      const stack: Array<[string, number]> = [];

      // 遍历字符串记录括号状态
      for (let i = 0; i < result.length; i++) {
        const c = result[i];

        for (const [open, close] of BRACKET_PAIRS) {
          if (c === open) {
            stack.push([open, i]); // 记录括号类型和位置
          } else if (c === close && stack.length > 0 && stack[stack.length - 1][0] === open) {
            stack.pop();
          }
        }
      }

      // 如果存在未闭合括号
      if (stack.length > 0) {
        const lastUnmatchedPos = stack[stack.length - 1][1];
        result = result.substring(0, lastUnmatchedPos).trimRight(); // 截断并移除末尾空格
        triggered = true;
      }
    }

    // 移除指定的尾部符号
    for (const sign of removeTailingSignList) {
      if (result.endsWith(sign)) {
        result = result.substring(0, result.length - sign.length).trimRight();
        triggered = true;
      }
    }

    // 没触发？退出循环
    if (!triggered) {
      break;
    }
  }

  return result;
}
